end-to-end test fails on CI occasionally

CI will catch the following error occasionally using `go test -race`.

    --- FAIL: Test_QueryEndToEnd (44.04s)
    --- FAIL: Test_QueryEndToEnd/selector_preserve_time.flux (0.01s)
        failed to run query: failed to compile query: type error 4:5-4:23: cannot unify object with semantic.Nature

It has only happened to me twice but both times it occurred on the above end-to-end test.

--- a/semantic/constraints.go
+++ b/semantic/constraints.go
@@ -96,7 +96,11 @@ func (v ConstraintGenerator) typeof(n Node) (PolyType, error) {
 		ftv := n.ExternType.freeVars(nil)
 		subst := make(Substitution, len(ftv))
 		for _, tv := range ftv {
-			subst[tv] = v.cs.f.Fresh()
+			f := v.cs.f.Fresh()
+			for ftv.contains(f) {
+				f = v.cs.f.Fresh()
+			}
+			subst[tv] = f
 		}
 		t := subst.ApplyType(n.ExternType)
 		// Check if this type knows about its kind constraints

--- a/semantic/inference_test.go
+++ b/semantic/inference_test.go
@@ -448,6 +448,65 @@ identity(x:identity)(x:2)
 			},
 		},
 		{
+			name: "extern type variables",
+			node: &semantic.Extern{
+				Assignments: []*semantic.ExternalVariableAssignment{
+					{
+						Identifier: &semantic.Identifier{Name: "f"},
+						ExternType: semantic.NewFunctionPolyType(
+							semantic.FunctionPolySignature{
+								Return: semantic.Tvar(3),
+							},
+						),
+					},
+					{
+						Identifier: &semantic.Identifier{Name: "g"},
+						ExternType: semantic.NewFunctionPolyType(
+							semantic.FunctionPolySignature{
+								Return: semantic.Tvar(5),
+							},
+						),
+					},
+				},
+				Block: &semantic.ExternBlock{
+					Node: &semantic.Program{
+						Body: []semantic.Statement{
+							&semantic.NativeVariableAssignment{
+								Identifier: &semantic.Identifier{Name: "a"},
+								Init:       &semantic.IdentifierExpression{Name: "f"},
+							},
+							&semantic.NativeVariableAssignment{
+								Identifier: &semantic.Identifier{Name: "b"},
+								Init:       &semantic.IdentifierExpression{Name: "g"},
+							},
+						},
+					},
+				},
+			},
+			solution: &solutionVisitor{
+				f: func(node semantic.Node) semantic.PolyType {
+					f := semantic.NewFunctionPolyType(
+						semantic.FunctionPolySignature{
+							Return: semantic.Tvar(7),
+						})
+					g := semantic.NewFunctionPolyType(
+						semantic.FunctionPolySignature{
+							Return: semantic.Tvar(8),
+						})
+					switch n := node.(type) {
+					case *semantic.IdentifierExpression:
+						switch n.Name {
+						case "f":
+							return f
+						case "g":
+							return g
+						}
+					}
+					return nil
+				},
+			},
+		},
+		{
 			name: "nested functions",
 			script: `
 (r) => {

